Firmware/MotorControl/pwm_input.cpp

```diff
@@ -0,0 +1,423 @@
+
+    endpoint->set_from_float((dir_pin_state) ? channel_value[gpio_index] : -channel_value[gpio_index]);
+
+    diff_steering_mixer_update(gpio_index+1); // User-facing GPIO numbers are 1-indexed
+}
+
+
+/**
+ * @brief Interpret RC or duty-cycle PWM into a value between min and max according to the pwm_mappings settings.
+ * 
+ * @param gpio_index The zero-indexed GPIO pin.  GPIO pin 1 is index 0.
+ * @param high_time Time between LtoH and HtoL transitions
+ * @param period Time between LtoH transitions
+ * @return true When a valid pulse is interpreted or no endpoint is set
+ * @return false When a pulse is invalid
+ */
+bool handle_pulse(int32_t gpio_index, Endpoint* endpoint, uint32_t high_time, uint32_t period) {
+    float ratio;
+    switch(board_config.pwm_mappings[gpio_index].pwm_type) {
+        case PWM_TYPE_RC: {
+            if (high_time < RC_MIN_LEGAL_HIGH_TIME || high_time > RC_MAX_LEGAL_HIGH_TIME)
+                return false;
+            ratio = (float)(high_time - RC_MIN_HIGH_TIME) / (float)(RC_MAX_HIGH_TIME - RC_MIN_HIGH_TIME);
+            ratio = (ratio > 1.0f) ? 1.0f : ((ratio < 0.0f) ? 0.0f : ratio); // Limit to [0.0, 1.0]
+        } break;
+        case PWM_TYPE_DUTY_CYCLE: {
+            if (high_time > period)
+                return false;
+            ratio = (float)high_time / (float)period;
+        } break;
+        default: {
+            return true;
+        } break;
+    }
+
+    if (board_config.pwm_mappings[gpio_index].enable_deadband) {
+        ratio = add_deadband(ratio);
+    }
+
+    set_channel_output(endpoint, gpio_index, ratio);
+
+    return true;
+}
+
+// In this function we need to calculate the duty cycle of the incoming PWM signal so that it can be translated into a control input
+
+// Three types of pulse-width modulation (PWM) are possible:
+// 1. The pulse center may be fixed in the center of the time window and both edges of the pulse moved to compress or expand the width.
+// 2. The lead edge can be held at the lead edge of the window and the tail edge modulated.
+// 3. The tail edge can be fixed and the lead edge modulated.
+
+// Is there a good way to check the period of all 3 types?  If not, type 2 should be the most common in this situation.
+// Type 2 is also the RC PWM signal type, so this implementation will work for both.
+
+// How can we verify that a pulse is a valid signal? RC PWM has this built-in, where a pulse needs to be within a certain length, and the period should be around
+// 20ms, but duty-cycle PWM has no such restrictions.  Right after boot, the old code waits for several HtoL transitions before it considers the input valid, but
+// shouldn't this be done every time a signal goes inactive?
+// We could compare the period of each pulse to the previous one, and require one or more consecutive matches for the pulse to be considered valid.  This would
+// work for RC PWM too, and would replace the previous validation method.
```

Comment:
If we ask the user to supply a nominal frequency, this becomes pretty straight forward: we can just check if the period is within a certain tolerance of the nominal period.
Regarding delay after start-up/bad signal, do we really need that? I don't remember/see such a thing in the old code.